src/Microsoft.Extensions.FileProviders.Physical/PhysicalFilesWatcher.cs

```diff
@@ -230,20 +240,59 @@ private void ReportChangeForMatchedEntries(string path)
         {
             path = NormalizePath(path);
 
+            var matched = false;
             ChangeTokenInfo matchInfo;
-            if (_matchInfoCache.TryRemove(path, out matchInfo))
+            if (_filePathTokenLookup.TryRemove(path, out matchInfo))
             {
                 CancelToken(matchInfo);
+                matched = true;
+            }
+
+            foreach (var wildCardEntry in _wildcardTokenLookup)
+            {
+                var matchResult = wildCardEntry.Value.Matcher.Match(path);
+                if (matchResult.HasMatches &&
+                    _wildcardTokenLookup.TryRemove(wildCardEntry.Key, out matchInfo))
+                {
+                    CancelToken(matchInfo);
+                    matched = true;
+                }
+            }
+
+            if (matched)
+            {
+                TryDisableFileSystemWatcher();
+            }
+        }
 
-                if (_matchInfoCache.Count == 0)
+        private void TryDisableFileSystemWatcher()
+        {
+            if (_filePathTokenLookup.IsEmpty && _wildcardTokenLookup.IsEmpty)
```

Comment:
@davidfowl told me not to use the if-lock-if pattern because it can break on hardware such as ARM